File: packages/flags-java/src/test/java/com/signakit/flags/AudienceMatcherTest.java
package com.signakit.flags;

import com.signakit.flags.audience.AudienceMatcher;
import com.signakit.flags.config.ConfigModels.AudienceCondition;
import com.signakit.flags.config.ConfigModels.AudienceMatchType;
import com.signakit.flags.config.ConfigModels.ConfigRuleAudience;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class AudienceMatcherTest {

    private static AudienceCondition cond(String attr, String op, Object val) {
        AudienceCondition c = new AudienceCondition();
        c.attribute = attr;
        c.operator = op;
        c.value = val;
        return c;
    }

    private static ConfigRuleAudience aud(AudienceCondition... conditions) {
        ConfigRuleAudience a = new ConfigRuleAudience();
        a.conditions = List.of(conditions);
        return a;
    }

    // -------------------------------------------------------------------------
    // equals / not_equals
    // -------------------------------------------------------------------------

    @Test
    void equals_matchesIdenticalStringValues() {
        assertTrue(AudienceMatcher.matchesCondition(
                cond("plan", "equals", "premium"), Map.of("plan", "premium")));
    }

    @Test
    void equals_rejectsDifferentStringValues() {
        assertFalse(AudienceMatcher.matchesCondition(
                cond("plan", "equals", "premium"), Map.of("plan", "free")));
    }

    @Test
    void equals_matchesBooleanValues() {
        assertTrue(AudienceMatcher.matchesCondition(
                cond("verified", "equals", true), Map.of("verified", true)));
        assertFalse(AudienceMatcher.matchesCondition(
                cond("verified", "equals", true), Map.of("verified", false)));
    }

    @Test
    void notEquals_matchesWhenValuesDiffer() {
        assertTrue(AudienceMatcher.matchesCondition(
                cond("plan", "not_equals", "premium"), Map.of("plan", "free")));
        assertFalse(AudienceMatcher.matchesCondition(
                cond("plan", "not_equals", "premium"), Map.of("plan", "premium")));
    }

    @Test
    void returnsFalseWhenAttributeIsMissing() {
        assertFalse(AudienceMatcher.matchesCondition(
                cond("plan", "equals", "premium"), Map.of()));
        assertFalse(AudienceMatcher.matchesCondition(
                cond("plan", "equals", "premium"), null));
    }

    // -------------------------------------------------------------------------
    // Numeric comparisons
    // -------------------------------------------------------------------------

    @Test
    void greaterThan_trueWhenUserValueExceedsThreshold() {
        assertTrue(AudienceMatcher.matchesCondition(
                cond("age", "greater_than", 18), Map.of("age", 25)));
        assertFalse(AudienceMatcher.matchesCondition(
                cond("age", "greater_than", 18), Map.of("age", 18)));
        assertFalse(AudienceMatcher.matchesCondition(
                cond("age", "greater_than", 18), Map.of("age", 10)));
    }

    @Test
    void lessThan_trueWhenUserValueIsBelowThreshold() {
        assertTrue(AudienceMatcher.matchesCondition(
                cond("age", "less_than", 18), Map.of("age", 10)));
        assertFalse(AudienceMatcher.matchesCondition(
                cond("age", "less_than", 18), Map.of("age", 18)));
        assertFalse(AudienceMatcher.matchesCondition(
                cond("age", "less_than", 18), Map.of("age", 25)));
    }

    @Test
    void greaterThanOrEqual_inclusive() {
        assertTrue(AudienceMatcher.matchesCondition(
                cond("age", "greater_than_or_equals", 18), Map.of("age", 18)));
        assertFalse(AudienceMatcher.matchesCondition(
                cond("age", "greater_than_or_equals", 18), Map.of("age", 17)));
    }

    @Test
    void lessThanOrEqual_inclusive() {
        assertTrue(AudienceMatcher.matchesCondition(
                cond("age", "less_than_or_equals", 18), Map.of("age", 18)));
        assertFalse(AudienceMatcher.matchesCondition(
                cond("age", "less_than_or_equals", 18), Map.of("age", 19)));
    }

    @Test
    void numericOperators_returnFalseOnStringAttributeValues() {
        assertFalse(AudienceMatcher.matchesCondition(
                cond("age", "greater_than", 18), Map.of("age", "25")));
    }

    // -------------------------------------------------------------------------
    // in / not_in
    // -------------------------------------------------------------------------

    @Test
    void in_trueWhenUserValueIsInTheList() {
        assertTrue(AudienceMatcher.matchesCondition(
                cond("country", "in", List.of("US", "CA", "GB")), Map.of("country", "US")));
        assertFalse(AudienceMatcher.matchesCondition(
                cond("country", "in", List.of("US", "CA", "GB")), Map.of("country", "DE")));
    }

    @Test
    void in_falseWhenValueIsNotAList() {
        assertFalse(AudienceMatcher.matchesCondition(
                cond("country", "in", "US"), Map.of("country", "US")));
    }

    @Test
    void notIn_trueWhenUserValueIsAbsentFromList() {
        assertTrue(AudienceMatcher.matchesCondition(
                cond("country", "not_in", List.of("US", "CA")), Map.of("country", "DE")));
        assertFalse(AudienceMatcher.matchesCondition(
                cond("country", "not_in", List.of("US", "CA")), Map.of("country", "US")));
    }

    @Test
    void notIn_trueWhenValueIsNotAList() {
        // Java: non-list value → vacuously true (mirrors JS/Python/Go SDK behaviour)
        assertTrue(AudienceMatcher.matchesCondition(
                cond("country", "not_in", "US"), Map.of("country", "US")));
    }

    // -------------------------------------------------------------------------
    // contains / not_contains
    // -------------------------------------------------------------------------

    @Test
    void contains_trueWhenStringIncludesSubstring() {
        assertTrue(AudienceMatcher.matchesCondition(
                cond("email", "contains", "@acme"), Map.of("email", "bob@acme.com")));
        assertFalse(AudienceMatcher.matchesCondition(
                cond("email", "contains", "@acme"), Map.of("email", "bob@gmail.com")));
    }

    @Test
    void contains_trueWhenStringArrayIncludesValue() {
        assertTrue(AudienceMatcher.matchesCondition(
                cond("tags", "contains", "beta"),
                Map.of("tags", List.of("alpha", "beta", "gamma"))));
        assertFalse(AudienceMatcher.matchesCondition(
                cond("tags", "contains", "beta"),
                Map.of("tags", List.of("alpha", "gamma"))));
    }

    @Test
    void notContains_trueWhenStringDoesNotIncludeSubstring() {
        assertTrue(AudienceMatcher.matchesCondition(
                cond("email", "not_contains", "@acme"), Map.of("email", "bob@gmail.com")));
        assertFalse(AudienceMatcher.matchesCondition(
                cond("email", "not_contains", "@acme"), Map.of("email", "bob@acme.com")));
    }

    @Test
    void notContains_trueWhenArrayDoesNotIncludeValue() {
        assertTrue(AudienceMatcher.matchesCondition(
                cond("tags", "not_contains", "beta"),
                Map.of("tags", List.of("alpha", "gamma"))));
        assertFalse(AudienceMatcher.matchesCondition(
                cond("tags", "not_contains", "beta"),
                Map.of("tags", List.of("alpha", "beta"))));
    }

    @Test
    void notContains_trueWhenTypesDoNotMatch() {
        // Java: numeric attribute with string needle → vacuously true (mirrors JS/Python/Go SDK)
        assertTrue(AudienceMatcher.matchesCondition(
                cond("score", "not_contains", "high"), Map.of("score", 42.0)));
    }

    // -------------------------------------------------------------------------
    // matchesAudience (via single-element list to matchesAudiences)
    // -------------------------------------------------------------------------

    @Test
    void matchesAudience_trueWhenAllConditionsMatch() {
        ConfigRuleAudience a = aud(
                cond("plan", "equals", "premium"),
                cond("age", "greater_than_or_equals", 18));
        assertTrue(AudienceMatcher.matchesAudiences(
                List.of(a), AudienceMatchType.ALL,
                Map.of("plan", "premium", "age", 25)));
    }

    @Test
    void matchesAudience_falseWhenAnyConditionFails() {
        ConfigRuleAudience a = aud(
                cond("plan", "equals", "premium"),
                cond("age", "greater_than_or_equals", 18));
        assertFalse(AudienceMatcher.matchesAudiences(
                List.of(a), AudienceMatchType.ALL,
                Map.of("plan", "premium", "age", 16)));
    }

    @Test
    void matchesAudience_trueForEmptyConditions() {
        ConfigRuleAudience a = aud();
        assertTrue(AudienceMatcher.matchesAudiences(
                List.of(a), AudienceMatchType.ALL, Map.of()));
    }

    // -------------------------------------------------------------------------
    // matchesAudiences
    // -------------------------------------------------------------------------

    @Test
    void matchesAudiences_trueWhenNullOrEmpty() {
        assertTrue(AudienceMatcher.matchesAudiences(null,      AudienceMatchType.ANY, Map.of()));
        assertTrue(AudienceMatcher.matchesAudiences(List.of(), AudienceMatchType.ANY, Map.of("plan", "premium")));
        assertTrue(AudienceMatcher.matchesAudiences(List.of(), AudienceMatchType.ALL, Map.of("plan", "premium")));
    }

    @Test
    void any_trueWhenAtLeastOneAudienceMatches() {
        List<ConfigRuleAudience> audiences = List.of(
                aud(cond("plan", "equals", "premium")),
                aud(cond("plan", "equals", "enterprise")));
        assertTrue(AudienceMatcher.matchesAudiences(
                audiences, AudienceMatchType.ANY, Map.of("plan", "premium")));
        assertFalse(AudienceMatcher.matchesAudiences(
                audiences, AudienceMatchType.ANY, Map.of("plan", "free")));
    }

    @Test
    void all_trueOnlyWhenEveryAudienceMatches() {
        List<ConfigRuleAudience> audiences = List.of(
                aud(cond("plan", "equals", "premium")),
                aud(cond("verified", "equals", true)));
        assertTrue(AudienceMatcher.matchesAudiences(
                audiences, AudienceMatchType.ALL,
                Map.of("plan", "premium", "verified", true)));
        assertFalse(AudienceMatcher.matchesAudiences(
                audiences, AudienceMatchType.ALL,
                Map.of("plan", "premium", "verified", false)));
    }

    @Test
    void defaultsToAllLogicWhenMatchTypeIsNull() {
        // null AudienceMatchType falls through to the "all" (AND) default.
        List<ConfigRuleAudience> audiences = List.of(
                aud(cond("plan", "equals", "premium")),
                aud(cond("verified", "equals", true)));
        assertTrue(AudienceMatcher.matchesAudiences(
                audiences, null,
                Map.of("plan", "premium", "verified", true)));
        assertFalse(AudienceMatcher.matchesAudiences(
                audiences, null,
                Map.of("plan", "premium", "verified", false)));
    }
}

File: packages/flags-php/tests/AudienceMatcherTest.php
<?php

declare(strict_types=1);

namespace SignaKit\FlagsPhp\Tests;

use PHPUnit\Framework\TestCase;
use SignaKit\FlagsPhp\AudienceMatcher;

final class AudienceMatcherTest extends TestCase
{
    // -------------------------------------------------------------------------
    // matchesCondition — equals / not_equals
    // -------------------------------------------------------------------------

    public function test_equals_matches_identical_string_values(): void
    {
        $cond = ['attribute' => 'plan', 'operator' => 'equals', 'value' => 'premium'];
        $this->assertTrue(AudienceMatcher::matchesCondition($cond, ['plan' => 'premium']));
    }

    public function test_equals_rejects_different_string_values(): void
    {
        $cond = ['attribute' => 'plan', 'operator' => 'equals', 'value' => 'premium'];
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, ['plan' => 'free']));
    }

    public function test_equals_matches_boolean_values(): void
    {
        $cond = ['attribute' => 'verified', 'operator' => 'equals', 'value' => true];
        $this->assertTrue(AudienceMatcher::matchesCondition($cond, ['verified' => true]));
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, ['verified' => false]));
    }

    public function test_not_equals_matches_when_values_differ(): void
    {
        $cond = ['attribute' => 'plan', 'operator' => 'not_equals', 'value' => 'premium'];
        $this->assertTrue(AudienceMatcher::matchesCondition($cond, ['plan' => 'free']));
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, ['plan' => 'premium']));
    }

    public function test_returns_false_when_attribute_is_missing(): void
    {
        $cond = ['attribute' => 'plan', 'operator' => 'equals', 'value' => 'premium'];
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, []));
    }

    // -------------------------------------------------------------------------
    // matchesCondition — numeric comparisons
    // -------------------------------------------------------------------------

    public function test_greater_than_true_when_user_value_exceeds_threshold(): void
    {
        $cond = ['attribute' => 'age', 'operator' => 'greater_than', 'value' => 18];
        $this->assertTrue(AudienceMatcher::matchesCondition($cond, ['age' => 25]));
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, ['age' => 18]));
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, ['age' => 10]));
    }

    public function test_less_than_true_when_user_value_is_below_threshold(): void
    {
        $cond = ['attribute' => 'age', 'operator' => 'less_than', 'value' => 18];
        $this->assertTrue(AudienceMatcher::matchesCondition($cond, ['age' => 10]));
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, ['age' => 18]));
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, ['age' => 25]));
    }

    public function test_greater_than_or_equals_inclusive(): void
    {
        $cond = ['attribute' => 'age', 'operator' => 'greater_than_or_equals', 'value' => 18];
        $this->assertTrue(AudienceMatcher::matchesCondition($cond, ['age' => 18]));
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, ['age' => 17]));
    }

    public function test_less_than_or_equals_inclusive(): void
    {
        $cond = ['attribute' => 'age', 'operator' => 'less_than_or_equals', 'value' => 18];
        $this->assertTrue(AudienceMatcher::matchesCondition($cond, ['age' => 18]));
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, ['age' => 19]));
    }

    public function test_numeric_operators_return_false_on_non_numeric_attribute_values(): void
    {
        $cond = ['attribute' => 'age', 'operator' => 'greater_than', 'value' => 18];
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, ['age' => 'twenty-five']));
    }

    // -------------------------------------------------------------------------
    // matchesCondition — in / not_in
    // -------------------------------------------------------------------------

    public function test_in_true_when_user_value_is_in_the_list(): void
    {
        $cond = ['attribute' => 'country', 'operator' => 'in', 'value' => ['US', 'CA', 'GB']];
        $this->assertTrue(AudienceMatcher::matchesCondition($cond, ['country' => 'US']));
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, ['country' => 'DE']));
    }

    public function test_in_false_when_value_is_not_an_array(): void
    {
        $cond = ['attribute' => 'country', 'operator' => 'in', 'value' => 'US'];
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, ['country' => 'US']));
    }

    public function test_not_in_true_when_user_value_is_absent_from_the_list(): void
    {
        $cond = ['attribute' => 'country', 'operator' => 'not_in', 'value' => ['US', 'CA']];
        $this->assertTrue(AudienceMatcher::matchesCondition($cond, ['country' => 'DE']));
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, ['country' => 'US']));
    }

    public function test_not_in_false_when_value_is_not_an_array(): void
    {
        // PHP: is_array($expected) guard makes this false (not vacuously true like JS)
        $cond = ['attribute' => 'country', 'operator' => 'not_in', 'value' => 'US'];
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, ['country' => 'US']));
    }

    // -------------------------------------------------------------------------
    // matchesCondition — contains / not_contains
    // -------------------------------------------------------------------------

    public function test_contains_true_when_string_includes_substring(): void
    {
        $cond = ['attribute' => 'email', 'operator' => 'contains', 'value' => '@acme'];
        $this->assertTrue(AudienceMatcher::matchesCondition($cond, ['email' => 'bob@acme.com']));
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, ['email' => 'bob@gmail.com']));
    }

    public function test_not_contains_true_when_string_does_not_include_substring(): void
    {
        $cond = ['attribute' => 'email', 'operator' => 'not_contains', 'value' => '@acme'];
        $this->assertTrue(AudienceMatcher::matchesCondition($cond, ['email' => 'bob@gmail.com']));
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, ['email' => 'bob@acme.com']));
    }

    public function test_not_contains_false_when_types_do_not_match(): void
    {
        // PHP: is_string($actual) && is_string($expected) guard — non-string attribute returns false
        $cond = ['attribute' => 'score', 'operator' => 'not_contains', 'value' => 'high'];
        $this->assertFalse(AudienceMatcher::matchesCondition($cond, ['score' => 42]));
    }

    // -------------------------------------------------------------------------
    // matchesAudiences
    // -------------------------------------------------------------------------

    public function test_returns_true_for_empty_audiences(): void
    {
        $this->assertTrue(AudienceMatcher::matchesAudiences([], 'any', []));
        $this->assertTrue(AudienceMatcher::matchesAudiences([], 'all', []));
    }

    public function test_any_returns_true_when_at_least_one_audience_matches(): void
    {
        $audiences = [
            ['conditions' => [['attribute' => 'plan', 'operator' => 'equals', 'value' => 'premium']]],
            ['conditions' => [['attribute' => 'plan', 'operator' => 'equals', 'value' => 'enterprise']]],
        ];
        $this->assertTrue(AudienceMatcher::matchesAudiences($audiences, 'any', ['plan' => 'premium']));
        $this->assertFalse(AudienceMatcher::matchesAudiences($audiences, 'any', ['plan' => 'free']));
    }

    public function test_all_returns_true_only_when_every_audience_matches(): void
    {
        $audiences = [
            ['conditions' => [['attribute' => 'plan', 'operator' => 'equals', 'value' => 'premium']]],
            ['conditions' => [['attribute' => 'verified', 'operator' => 'equals', 'value' => true]]],
        ];
        $this->assertTrue(AudienceMatcher::matchesAudiences($audiences, 'all', ['plan' => 'premium', 'verified' => true]));
        $this->assertFalse(AudienceMatcher::matchesAudiences($audiences, 'all', ['plan' => 'premium', 'verified' => false]));
    }

    public function test_all_conditions_in_a_single_audience_must_pass(): void
    {
        $audiences = [
            [
                'conditions' => [
                    ['attribute' => 'plan', 'operator' => 'equals', 'value' => 'premium'],
                    ['attribute' => 'age', 'operator' => 'greater_than_or_equals', 'value' => 18],
                ],
            ],
        ];
        $this->assertTrue(AudienceMatcher::matchesAudiences($audiences, 'any', ['plan' => 'premium', 'age' => 25]));
        $this->assertFalse(AudienceMatcher::matchesAudiences($audiences, 'any', ['plan' => 'premium', 'age' => 16]));
    }
}
